cylc/uiserver/websockets/tornado.py

```diff
@@ -0,0 +1,210 @@
+
+
+class TornadoConnectionContext(BaseConnectionContext):
+    async def receive(self):
+        try:
+            msg = await self.ws.recv()
+            return msg
+        except WebSocketClosedError:
+            raise ConnectionClosedException()
+
+    async def send(self, data):
+        if self.closed:
+            return
+        await self.ws.write_message(data)
+
+    @property
+    def closed(self):
+        return self.ws.close_code is not None
+
+    async def close(self, code):
+        await self.ws.close(code)
+
+
+class TornadoSubscriptionServer(BaseSubscriptionServer):
+    def __init__(self, schema, keep_alive=True, loop=None):
+        self.loop = loop
+        super().__init__(schema, keep_alive)
+
+    def get_graphql_params(self, *args, **kwargs):
+        params = super(TornadoSubscriptionServer,
+                       self).get_graphql_params(*args, **kwargs)
+        return dict(params, return_promise=True, executor=AsyncioExecutor(loop=self.loop))
+
+    async def _handle(self, ws, request_context):
+        connection_context = TornadoConnectionContext(ws, request_context)
+        await self.on_open(connection_context)
+        pending = set()
+        while True:
+            try:
+                if connection_context.closed:
+                    raise ConnectionClosedException()
+                message = await connection_context.receive()
+            except ConnectionClosedException:
+                break
+            finally:
+                if pending:
+                    (_, pending) = await wait(pending, timeout=0, loop=self.loop)
+
+            task = ensure_future(
+                self.on_message(connection_context, message), loop=self.loop)
+            pending.add(task)
+
+        self.on_close(connection_context)
+        for task in pending:
+            task.cancel()
+
+    async def handle(self, ws, request_context=None):
+        await shield(self._handle(ws, request_context), loop=self.loop)
+
+    async def on_open(self, connection_context):
+        pass
+
+    def on_close(self, connection_context):
+        remove_operations = list(connection_context.operations.keys())
+        for op_id in remove_operations:
+            self.unsubscribe(connection_context, op_id)
+
+    async def on_connect(self, connection_context, payload):
+        pass
+
+    async def on_connection_init(self, connection_context, op_id, payload):
+        try:
+            await self.on_connect(connection_context, payload)
+            await self.send_message(connection_context, op_type=GQL_CONNECTION_ACK)
+        except Exception as e:
+            await self.send_error(connection_context, op_id, e, GQL_CONNECTION_ERROR)
+            await connection_context.close(1011)
+
+    def execute(self, request_context, params):
+        params['context_value'] = request_context
+        return super().execute(request_context, params)
+
+    async def send_execution_result(self, connection_context, op_id,
```

Comment:
@dwsutherland the solution you linked was added to the cylc-uiserver code from Tornado PR, not to cylc-flow :+1:

Reply:
Nice 👍  ... Should really be part of the code base, as it overwrites:
https://github.com/graphql-python/graphql-ws/blob/523dd1516d8709149a475602042d6a3c5eab5a3d/graphql_ws/base.py#L113-L115

Although it would have to handle more than just asyncio I suppose (probably why)

Reply:
But it does give users a nice feature of being able to re-use the Query schema (even if partially). Otherwise we would have to simply duplicate all those fields that have resolvers :open_mouth: